flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/AdaptedRestartPipelinedRegionStrategyNGFailoverTest.java

```diff
@@ -0,0 +1,513 @@
+
+		final Iterator<ExecutionVertex> vertexIterator = eg.getAllExecutionVertices().iterator();
+		final ExecutionVertex ev11 = vertexIterator.next();
+		final ExecutionVertex ev12 = vertexIterator.next();
+		final ExecutionVertex ev21 = vertexIterator.next();
+		final ExecutionVertex ev22 = vertexIterator.next();
+
+		// trigger task failure of ev11
+		// regions {ev11}, {ev21}, {ev22} should be affected
+		ev11.getCurrentExecutionAttempt().fail(new Exception("Test Exception"));
+		manualMainThreadExecutor.triggerAll();
+
+		// verify vertex states
+		// only vertices with consumable inputs can be scheduled
+		assertVertexInState(ExecutionState.DEPLOYING, ev11);
+		assertVertexInState(ExecutionState.DEPLOYING, ev12);
+		assertVertexInState(ExecutionState.CREATED, ev21);
+		assertVertexInState(ExecutionState.CREATED, ev22);
+
+		// verify attempt number
+		assertEquals(1, ev11.getCurrentExecutionAttempt().getAttemptNumber());
+		assertEquals(0, ev12.getCurrentExecutionAttempt().getAttemptNumber());
+		assertEquals(1, ev21.getCurrentExecutionAttempt().getAttemptNumber());
+		assertEquals(1, ev22.getCurrentExecutionAttempt().getAttemptNumber());
+	}
+
+	/**
+	 * Tests that the failure is properly propagated to underlying strategy
+	 * to calculate tasks to restart.
+	 * <pre>
+	 *     (v11) -+-> (v21)
+	 *            x
+	 *     (v12) -+-> (v22)
+	 *
+	 *            ^
+	 *            |
+	 *        (blocking)
+	 * </pre>
+	 */
+	@Test
+	public void testFailurePropagationToUnderlyingStrategy() throws Exception {
+		final JobGraph jobGraph = createBatchJobGraph();
+		final ExecutionGraph eg = createExecutionGraph(jobGraph);
+
+		final TestAdaptedRestartPipelinedRegionStrategyNG failoverStrategy =
+			(TestAdaptedRestartPipelinedRegionStrategyNG) eg.getFailoverStrategy();
+
+		final Iterator<ExecutionVertex> vertexIterator = eg.getAllExecutionVertices().iterator();
+		final ExecutionVertex ev11 = vertexIterator.next();
+		final ExecutionVertex ev12 = vertexIterator.next();
+		final ExecutionVertex ev21 = vertexIterator.next();
+
+		// trigger downstream regions to schedule
+		componentMainThreadExecutor.execute(() -> {
+			// finish upstream regions to trigger scheduling of downstream regions
+			ev11.getCurrentExecutionAttempt().markFinished();
+			ev12.getCurrentExecutionAttempt().markFinished();
+		});
+
+		// trigger task failure of ev21 on consuming data from ev11
+		Exception taskFailureCause = new PartitionConnectionException(
+			new ResultPartitionID(
+				ev11.getProducedPartitions().keySet().iterator().next(),
+				ev11.getCurrentExecutionAttempt().getAttemptId()),
+			new Exception("Test failure"));
+		ev21.getCurrentExecutionAttempt().fail(taskFailureCause);
+		manualMainThreadExecutor.triggerAll();
+
+		assertThat(failoverStrategy.getLastTasksToCancel(), containsInAnyOrder(
+			new ExecutionVertexID(ev11.getJobvertexId(), 0),
+			new ExecutionVertexID(ev21.getJobvertexId(), 0),
+			new ExecutionVertexID(ev21.getJobvertexId(), 1)));
```

Comment:
would be easier to understand if this were `ev22.getJobvertexId`

Reply:
true, I am waiting for `ev22.getID()`

Reply:
Fixed. Commit cherry picked from #8804.